### ISSUE TITLE
Create Bundle item

### ISSUE THREAD

**Issue by Eforen** [inventree/InvenTree#2624]
I may just not understand something but how do I create a bundled item that breaks down into 2 other items?

For example if my supplier only sells super glue in bundles of 2 how can i make it so that I have a super glue bundle part that can be tracked in inventory and then an individual superglue part that is created in the quantity 2 when a super glue bundle is broken down into 2 separate glue bottles by opening it?

**Comment by SchrodingersGat** [comment]
Not stale

**Comment by miggland** [comment]
Hasn't this been closed by https://github.com/inventree/InvenTree/pull/3644?

**Comment by SchrodingersGat** [comment]
@miggland no, separate issue. Pack size is about "number of items" but a bundle could be multiple *different* items

**Comment by miggland** [comment]
Pack size sounds what OP is requesting.

However, bundles are a great idea also. :+1:

**Comment by Eforen** [comment]
> Pack size sounds what OP is requesting.
> 
> However, bundles are a great idea also. :+1: 

Stack size is what I was requesting but bundle would be really useful say for example it's cheaper to buy a pack of 10 nut and bolt pairs but you only need the nuts right now. That would be very useful.

**Comment by SchrodingersGat** [comment]
Not stale

**Comment by chris-cole-walker** [comment]
I am really interested in seeing this feature added to InvenTree so I am developing a simple implementation that works using existing features and no model changes to "expand" stock items of assemblies into new stock items based on the BOM.

The implementation is basically done and I just need to finalize testing before I can open a pull request for the feature.

If anybody has any input on what this feature should be called, it would be appreciated. Right now, I have called it "expand" to keep it short, but I think "disassemble" or any other term to imply it is breaking down the stock item would work. Attached is a screen shot of what it would look like, using the `fas fa-object-ungroup` icon in the menu.

<img width="435" alt="Expand Stock" src="https://github.com/inventree/InvenTree/assets/60531254/5c65a5ed-0d4f-4085-9fe7-29ffef6ce173">

**Comment by SchrodingersGat** [comment]
@chris-cole-walker looks great! I think "disassemble" makes more sense than "expand". What control does the user have over the process?

**Comment by chris-cole-walker** [comment]
@SchrodingersGat thanks! Yeah, I was also leaning towards something more descriptive like "disassemble" which is why I brought it up. I'll make that change before I move forward.

As of this moment, the user (assuming they have `role.stock.change`) really can only control the number of items that is disassembled (i.e. if you have 10 blue chairs do you disassemble 1 to get 4 legs or 10 to get 40 legs) and the location that all of the disassembled pieces are placed. Those were the two things I saw as easiest to implement using all existing forms and ideas. The user can also add a note, similar to any other stock adjustment, but I created two new StockHistoryCodes to explicitly track when stock was disassembled and created from disassembly.

I did play with adding an additional attribute to the part model that would indicate if it could be disassembled, but I felt it just added an unnecessary item to the new part form and was cleaner to just assume all assemblies could be disassembled. That being said, it would be an additional control for users with `role.part.add/change` and would be simple to add if you felt it was necessary.

**Comment by SchrodingersGat** [comment]
Nice, that all makes sense.

Another consideration - what happens if the item already has "installed items" (i.e. has been assembled and has tracked sub-assemblies inside it). Does your approach factor this case in?

**Comment by chris-cole-walker** [comment]
Short answer, sort of... I am addressing "installed items" by not allowing disassembly if they are present. 

Long answer...
At the moment, I have used nearly identical criteria as when checking if a merge is valid to check if disassembly is valid. Anything "installed" in another item, with "installed items", being built, allocated to a sales order, or assigned to a customer cannot be disassembled. I have also prevented serialized items from being disassembled because I don't know what would be the proper way to handle that: "forget" and allow the serial number to be assigned to a new part down the road or track it so it maintains its uniqueness and cannot be used even though the part is gone.

I see two main use cases for "disassembly": an item that is a variety or multi-pack that is better tracked as its pieces (i.e. the glue example that started this thread) and an assembly built in-house that is disassembled to return its pieces back to stock (i.e. a chair is disassembled so the legs can be reallocated to a table). That first use case was my initial motivation and simple as it will never have "installed items," consumable parts, or variants. The second use case is more complicated and could involve "installed items." I think that would be best handled by more granular controls such as being presented with a modified view of the base part's BOM and specifying which items are re-stocked and which items are lost/consumed.

One thing I did just think about about since you asked is that I currently don't check to see if a BOM item is marked as consumable or allowing variants and the disassembly just trusts that all items in the BOM are the explicit part listed and can be re-stocked. This means if you disassemble a "Blue Chair," you get 5 screws, 4 legs, and 0.125 liters of blue paint back in stock. While screws are marked as "consumable" in the demo database, it is feasible they could be removed and restocked. Paint can also be sanded off of a chair leg so it can be used elsewhere. However, 0.125 liters of paint can't magically be re-stocked so this falls under the second use case I mentioned above and needs more input from the user at some point before or during the disassembly process.

**Comment by SchrodingersGat** [comment]
@chris-cole-walker it is certainly complicated - which is why the process has not yet been implemented :)

Maybe the "disassemble" process should require input from the user - which items are put into stock, and which are "vanished"?

**Comment by chris-cole-walker** [comment]
@SchrodingersGat Ok, I'll see what I can do this weekend!

**Comment by SchrodingersGat** [comment]
not stale

**Comment by SchrodingersGat** [comment]
Not stale

**Comment by SchrodingersGat** [comment]
Not stale

**Comment by SchrodingersGat** [comment]
Not stale, just taking a *really* long time to get to! Still do not have a clean implementation pathway for this.

**Comment by bmalatest** [comment]
We would be happy to help fund this as it is coming up during our production use of InvenTree. Specifically that after stock is marked as consumed that it is no longer editable but as highlighted in the related issues and description, there are multiple reasons why you may need to update the perceived to be consumed stock to be available and with a higher quantity than after being consumed. What level of effort would be required for this? How can we help fund this in a reasonable amount of time to be implemented?

**Comment by SchrodingersGat** [comment]
@bmalatest can you expand on your use case here? I do not think it is necessarily the same as the original posted issue here, but maybe I am not understanding correctlyl.

**Comment by SchrodingersGat** [comment]
FYI - this feature is now incoming in https://github.com/inventree/InvenTree/pull/12310
